### PR TITLE
Requesting a code review of the VF2 Algorithm for ISSUE: 87

### PR DESCRIPTION
**Pull request by sracha4355**
The algorithm is tested and it works well. I still need to add better comments for the code and the tests, but I would like a initial code review for graph_isomorphism.h and graph_isomorpishm.tpp (please ignore the comments in graph_isomorphism.tpp for now). The VF2 algorithm in addition to comparing the structure of the graph for isomorphism also allows for semantic comparison of graphs, meaning the attributes of the nodes. However, this code only takes into account the structure of the graphs. @bobluppes I was wondering if we can push this working code and add on this functionality later on? Thank you, please be brutal with my code as is my first time really working with C++ outside of school. 
P.S how do I check code coverage?

### CODE REVIEW

**Review by joweich — COMMENTED** [review]
Hey @sracha4355, huge congrats to your very first outside-of-school C++ PR 🚀 You definitely chose a challenging one for this and did very well! 
Some early guardrails for increasing compliance with best practices and making your code more concise:
- In the header file, you should only have the declarations that the user needs to know about. Everything else should be hidden in the `.tpp`
- Your code is quite verbose and has some potential for refactoring. For example, repetitive tasks make a great case for extracting a function and increasing code readability. Your feeling for this will grow quickly 🙂
- You're not only testing the final `check_isomorphism` function, but also all the steps in between. This a reasonable for debugging, but might limit refactoring opportunities down the road if you're doing test-driven development
Let me know if there's any additional support I can provide!🙂 Looking forward to your next commits!

File: include/graaflib/algorithm/graph_isomorphism.h [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+#include <graaflib/graph.h>
+#include <vector>
+#include <unordered_map>
+#include <optional>
+#include <unordered_set>
+#include <array>
+
+namespace graaf::algorithm {
+
+// Type aliases
```
Comment (joweich) [inline include/graaflib/algorithm/graph_isomorphism.h:14]: Following naming conventions, you'd start your alias names with a lowercase letter. You could also make `Mapping` a bit more descriptive, e.g. `vertex_mapping`.

File: include/graaflib/algorithm/graph_isomorphism.h [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+ * @tparam V The vertex type of the graphs.
+ * @tparam E The edge type of the graphs.
+ * @tparam T The graph type of the graphs (directed or undirected).
+ * @param lhs The first graph to compare.
+ * @param rhs The second graph to compare.
+ * @return An optional containing a mapping from the vertices of lhs to those of rhs if isomorphic; otherwise std::nullopt.
+ */
+template <typename V, typename E, graph_type T>
+std::optional<Mapping>
+check_isomorphism(
```
Comment (joweich) [inline include/graaflib/algorithm/graph_isomorphism.h:38]: This is the only function users are supposed to call, correct? All the other definitions should be hidden in the `.tpp` file then.

File: include/graaflib/algorithm/graph_isomorphism.h [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+ * @tparam V The vertex type of the graphs.
+ * @tparam E The edge type of the graphs.
+ * @tparam T The graph type of the graphs (directed or undirected).
+ * @param lhs The first graph to compare.
+ * @param rhs The second graph to compare.
+ * @return vf2_state struct
+ */
+template <typename V, typename E, graph_type T>
+vf2_state * init_vf2_state(
+	const graph <V, E, T> & lhs,
```
Comment (joweich) [inline include/graaflib/algorithm/graph_isomorphism.h:124]:
```suggestion
	const graph <V, E, T>& lhs,
```
The convention in this project is dropping the whitespace in these cases

**Review by joweich — COMMENTED** [review]
@sracha4355 I added a few more prescriptive comments to the code to give you a more tangible idea of the potentials for improvement mentioned above 🙂

**Comment by sracha4355** [comment]
> Hey @sracha4355, huge congrats on your very first outside-of-school C++ PR 🚀 You definitely chose a challenging one for this and did very well! Some early guardrails for increasing compliance with best practices and making your code more concise:
> 
> * In the header file, you should only have the declarations that the user needs to know about. Everything else should be hidden in the `.tpp`
> * Your code is quite verbose and has some potential for refactoring. For example, repetitive tasks make a great case for extracting a function and increasing code readability. Your feeling for this will grow quickly 🙂
> * You're not only testing the final `check_isomorphism` function, but also all the steps in between. This a reasonable for debugging, but might limit refactoring opportunities down the road if you're doing test-driven development
> 
> Let me know if there's any additional support I can provide!🙂 Looking forward to your next commits!

Thanks @joweich, do you recommend removing the tests related to the helper functions? I agree that it was helpful for debugging. Sure, I can definitely refactor the code. For comments, how descriptive should they be, should I try to explain every aspect of how and why the algorithm works this way? Is it mandatory to also add comments to my tests?

File: include/graaflib/algorithm/graph_isomorphism.h (hunk 2 above)
Comment (sracha4355) [reply include/graaflib/algorithm/graph_isomorphism.h:38]: Do the other functions that will not be called by the user need comments explaining how they work?

File: include/graaflib/algorithm/graph_isomorphism.h (hunk 3 above)
Comment (sracha4355) [reply include/graaflib/algorithm/graph_isomorphism.h:124]: ok gotcha

**Comment by joweich** [comment]
> For comments, how descriptive should they be, should I try to explain every aspect of how and why the algorithm works this way? Is it mandatory to also add comments to my tests?

@sracha4355 the best code is code that does not need comments to be comprehensive for readers. In other words, if your code is self-explanatory, there's no need for any comments (apart from the interface documentation with the @-signs)

**Comment by sracha4355** [comment]
@joweich is the .tpp not subject to clang formatting? How can I check if my new changes follow clang format. Is there a tool to automatically format the code for me

**Comment by joweich** [comment]
> @joweich is the .tpp not subject to clang formatting?

Yes, it is. Codestyle concerns every sourcefile 🙂

> How can I check if my new changes follow clang format. Is there a tool to automatically format the code for me

Check [this guide](https://github.com/bobluppes/graaf/wiki/development-setup#formatting) for using clang format. Most IDEs have inbuilt support for this

**Pull request closed without merging**